### ISSUE TITLE
1.4.22 regression: inserted_primary_key is None after successful insert where primary key passed as None

### ISSUE THREAD

**Issue by gward** [sqlalchemy/sqlalchemy#7998]
### Describe the bug

I'm using SQLAlchemy core to insert a row on a simple table with two fields: integer primary key and name: specifically, calling `table.insert().values()` with all values for the new row. If I include the primary key in the arguments to `values()`, explicitly passing None, then the `inserted_primary_key` that I get back after the INSERT is None -- even though that doesn't match what's in the database.

This appears to be a regression -- it worked with SQLAlchemy 1.4.21, but stopped working with 1.4.22.

### To Reproduce

```python
import sqlalchemy as sa

metadata = sa.MetaData()

tbl_thing = sa.Table(
    'thing', metadata,
    sa.Column('id', sa.Integer, primary_key=True),
    sa.Column('name', sa.String),
)

url = 'sqlite:///test.sqlite/'
engine = sa.create_engine(url)
metadata.create_all(engine)

print('connecting to', url)
conn = engine.connect()

stmt = tbl_thing.insert().values(
    id=None,
    name='x')
print(str(stmt))
result = conn.execute(stmt)
print('result:', result)
print('rowcount:', result.rowcount)
assert result.rowcount == 1
print('inserted_primary_key:', result.inserted_primary_key)
assert result.inserted_primary_key[0] is not None, repr(result.inserted_primary_key)
result.close()

conn.close()
```

### Error

```
connecting to sqlite:///test.sqlite/
INSERT INTO thing (id, name) VALUES (:id, :name)
result: <sqlalchemy.engine.cursor.LegacyCursorResult object at 0x7f094687bfd0>
rowcount: 1
inserted_primary_key: (None,)
Traceback (most recent call last):
  File "/tmp/gml/return-null-primary-key.py", line 27, in <module>
    assert result.inserted_primary_key[0] is not None, repr(result.inserted_primary_key)
AssertionError: (None,)
``` 

### Versions

- OS:Ubuntu 22.04 LTS
- Python: 3.10.4 (Ubuntu package in a virtualenv)
- SQLAlchemy: 1.4.22 .. 1.4.36 
- Database: sqlite 3.37.2 (Ubuntu package)
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): stdlib sqlite3 (I assume!)

### Additional context

If I run my test program with SQLAlchemy 1.4.21, it works as expected:

```
connecting to sqlite:///test.sqlite/
INSERT INTO thing (id, name) VALUES (NULL, :name)
result: <sqlalchemy.engine.cursor.LegacyCursorResult object at 0x7feb921f6410>
rowcount: 1
inserted_primary_key: (3,)
```

Likewise, if I make a tiny patch to the test program, it works as expected:

```
@@ -16,7 +16,6 @@
 conn = engine.connect()
 
 stmt = tbl_thing.insert().values(
-    id=None,
     name='x')
 print(str(stmt))
 result = conn.execute(stmt)
```

That code works with SQLAlchemy 1.4.22 and 1.4.36. The original test script works with 1.4.21, but fails with 1.4.22 and 1.4.36.

**Comment by zzzeek** [comment]
hey there -

the use case here is completely undefined behavior and just happens to work on SQLite, and apparently MySQL also.   The query will fail under all versions on a database like PostgreSQL, where the explicit NULL for PK is honored by PostgreSQL itself:

```
sqlalchemy.exc.IntegrityError: (psycopg2.errors.NotNullViolation) null value in column "id" of relation "thing" violates not-null constraint
DETAIL:  Failing row contains (null, x).

[SQL: INSERT INTO thing (id, name) VALUES (NULL, %(name)s) RETURNING thing.id]
```

The correct way to allow the primary key to autoincrement is to omit it from the list of columns being inserted.

we can look into testing for explicit support of following along with databases that discard an explicit NULL to generate a PK despite a parameter being passed but this is a brand new use case.

**Comment by gward** [comment]
> the use case here is completely undefined behavior and just happens to work on SQLite

Yeah, I thought it might be something like that. Well, even if you don't add support for this, at least there's a searchable bug report with a workaround for the next person to stumble upon this!

**Comment by zzzeek** [comment]
I can restore it but it would be local to when insert().values() is used using an explicit check for None.   it wouldn't occur if you said connection.execute(insert, {"id": None}) as the SQL compiler does not consult the "None" value in that case.

**Comment by zzzeek** [comment]
nevermind I'll get that one also, will be more consistent

**Comment by zzzeek** [comment]
1.3 worked that way

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**explicitly fetch inserted pk for values(pkcol=None)** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3828

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_4** branch:

**explicitly fetch inserted pk for values(pkcol=None)** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3829

**Comment by zzzeek** [comment]
this one is a little risky and would be easier to support in 2.0 alone....are you able to work around this in 1.4?

**Comment by gward** [comment]
I've just implemented the workaround, and it's fine. If it makes my code a little more portable to databases other than SQLite, then all the better!

So, I'd be fine if this doesn't get fixed in SQLAlchemy 1.4.x, and I'd be fine if it never gets fixed.

**Comment by zzzeek** [comment]
OK.  I think we should have it because the database backend is creating the new value, but it's better if this is a 2.0 thing as it's kind of a big change.  thanks for the prompt feedback
